### PR TITLE
dxChart: Fix calculation of initial value axis range (T629067)

### PR DESCRIPTION
**Pull request by dxBeardedBear**

### CODE REVIEW

File: js/viz/translators/translator2d.js [hunk 1]
```diff
@@ -149,9 +150,19 @@ function isEqualRange(range) {
     const businessRange = that.getBusinessRange();
     const canvasOptions = getCanvasBounds(businessRange);
     const correctionPrecision = getEqualityCorrection(businessRange) / 100;
+    const comparingRange = range && new rangeModule.Range(range);
```
Comment (alexanderPolosatov) [inline js/viz/translators/translator2d.js]: range && - extra operation. It's exist in constructor of Range

File: js/viz/translators/translator2d.js [hunk 2]
```diff
@@ -149,9 +150,19 @@ function isEqualRange(range) {
...
 
-    return range && _abs(range.min - canvasOptions.rangeMin) <= correctionPrecision &&
-        _abs(range.max - canvasOptions.rangeMax) <= correctionPrecision;
+    if(businessRange.axisType === "discrete") {
+        return false;
+    }
+
+    if(comparingRange && businessRange.axisType === "logarithmic") {
+        comparingRange.min = comparingRange.min && vizUtils.getLog(comparingRange.min, businessRange.base);
```
Comment (alexanderPolosatov) [inline js/viz/translators/translator2d.js]: You can use defined 'getLog' variable instead vizUtils.getLog.

File: testing/tests/DevExpress.viz.core/translator2D.tests.js [hunk 3]
```diff
@@ -645,6 +645,58 @@ QUnit.test('Get canvas visible area', function(assert) {
...
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { max: 9000 }));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { min: 500, max: 7000 }));
+    assert.ok(translator.constructor.prototype.isEqualRange.call(translator, {}));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, null));
+});
+
+QUnit.test('Compare datetime business range', function(assert) {
+    var canvas = $.extend({}, canvasTemplate),
+        translator = new translator2DModule.Translator2D($.extend({}, datetimeRange), canvas, optionsHorizontal);
+
```
Comment (alexanderPolosatov) [inline testing/tests/DevExpress.viz.core/translator2D.tests.js:675]: Extra line

File: testing/tests/DevExpress.viz.core/translator2D.tests.js [hunk 4]
```diff
@@ -645,6 +645,58 @@ QUnit.test('Get canvas visible area', function(assert) {
...
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { min: 0, max: 90 }));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { min: 50, max: 70 }));
+    assert.ok(translator.constructor.prototype.isEqualRange.call(translator, {}));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, null));
+});
+
+QUnit.test('Compare logarithmic business range', function(assert) {
+    var canvas = $.extend({}, canvasTemplate),
+        translator = new translator2DModule.Translator2D($.extend({}, logarithmicRange), canvas, optionsHorizontal);
+
```
Comment (alexanderPolosatov) [inline testing/tests/DevExpress.viz.core/translator2D.tests.js:663]: Extra line

File: testing/tests/DevExpress.viz.core/translator2D.tests.js [hunk 5]
```diff
@@ -645,6 +645,58 @@ QUnit.test('Get canvas visible area', function(assert) {
...
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { min: new Date(2012, 9, 1), max: new Date(2012, 9, 1, 13) }));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, { min: new Date(2012, 9, 1, 10), max: new Date(2012, 9, 1, 22) }));
+    assert.ok(translator.constructor.prototype.isEqualRange.call(translator, {}));
+    assert.ok(!translator.constructor.prototype.isEqualRange.call(translator, null));
+});
+
+QUnit.test('Compare discrete business range', function(assert) {
+    var canvas = $.extend({}, canvasTemplate),
+        translator = new translator2DModule.Translator2D($.extend({}, discreteRange), canvas, optionsHorizontal);
+
```
Comment (alexanderPolosatov) [inline testing/tests/DevExpress.viz.core/translator2D.tests.js:687]: Extra line

File: testing/tests/DevExpress.viz.core/translator2D.tests.js [hunk 6]
```diff
@@ -645,6 +645,58 @@ QUnit.test('Get canvas visible area', function(assert) {
     assert.deepEqual(visibleArea, { min: 70, max: 580 });
 });
 
+QUnit.test('Compare numeric business range', function(assert) {
+    var canvas = $.extend({}, canvasTemplate),
+        translator = new translator2DModule.Translator2D($.extend({}, numericRange), canvas, optionsHorizontal);
+
```
Comment (alexanderPolosatov) [inline testing/tests/DevExpress.viz.core/translator2D.tests.js:651]: Extra line

File: js/viz/translators/translator2d.js (hunk 1 above)
Comment (alexanderPolosatov) [reply js/viz/translators/translator2d.js]: and in 'if(comparingRange' use 'range' instead for this checking

**Review by alexanderPolosatov — CHANGES_REQUESTED** [review]

File: js/viz/translators/translator2d.js (hunk 1 above)
Comment (dxBeardedBear) [reply js/viz/translators/translator2d.js]: new rangeModule.Range(range) returns object anyway

File: testing/tests/DevExpress.viz.core/translator2D.tests.js (hunk 6 above)
Comment (dxBeardedBear) [reply testing/tests/DevExpress.viz.core/translator2D.tests.js:651]: done

File: testing/tests/DevExpress.viz.core/translator2D.tests.js (hunk 5 above)
Comment (dxBeardedBear) [reply testing/tests/DevExpress.viz.core/translator2D.tests.js:687]: done

File: testing/tests/DevExpress.viz.core/translator2D.tests.js (hunk 4 above)
Comment (dxBeardedBear) [reply testing/tests/DevExpress.viz.core/translator2D.tests.js:663]: done

File: testing/tests/DevExpress.viz.core/translator2D.tests.js (hunk 3 above)
Comment (dxBeardedBear) [reply testing/tests/DevExpress.viz.core/translator2D.tests.js:675]: done

File: js/viz/translators/translator2d.js (hunk 2 above)
Comment (dxBeardedBear) [reply js/viz/translators/translator2d.js]: done

File: js/viz/translators/translator2d.js (hunk 1 above)
Comment (dxBeardedBear) [reply js/viz/translators/translator2d.js]: done

**Review by alexanderPolosatov — APPROVED** [review]

**Pull request merged**